pageserver/src/tenant.rs

```diff
@@ -1389,13 +1445,107 @@ impl Tenant {
             timeline
         };
 
+        // multiple concurrent requests could had "transitioned" a timeline to stopping state,
+        // because repeat transitions to same state are allowed. because the delete_timeline will
+        // need to synchronize and cannot be original semi-lockless algorithm for "upload
+        // indexpart" part, we create a single task to delete the timeline.
```

Comment:
> cannot be original semi-lockless algorithm for "upload indexpart"

I disagree, see changes in https://github.com/neondatabase/neon/pull/4147 

Regardless, for this PR, I think there's no need to mention any of this here.
Git history holds the past!

What's more important to mention is
- why cmpxchg of timeline state wasn't chosen
- that this solution doesn't aim to prevent the race with tenant detach

```suggestion
        // Prevent concurrent calls to this function.
        // The `set_state(Stopping)` function above doesn't cut it.
        // One would think we could turn it into a cmpxchg-like function
        // and bail out early if we're already Stopping.
        // But, there are two caveats with that:
        // 1. What if we fail with a non-permanent error below?
        //    To allow retries of the dele, we would need to a state other than Stopping.
        //    TimelineState is a dangerous mess already, don't want to add more variants to it.
        // 2. This function is not the only one that can set TimelineState to Stopping.
        //    There's also tenant detach, and pageserver shutdown.
        //    Tenant detach also does the same thing as this function, i.e., deleting timelie dir.
        //    We'd need to implement similar cmpxchg in all those places to get it correct.
        //
        // So, we constrain ourselves to preventing concurrent timeline_deletes.
        // While leaving the other problems pointed out above for the future.
```

Reply:
Did we at some point have a version with anything atomic (cmpxchg)?

Reply:
I did, and discarded it, before publishing it for review

Reply:
I don't want to each PR to add a wall of text as it lowers the overall signal to noise ratio.

I'll try to come up something easily readable which doesn't refer to prev iterations.

> I did, and discarded it, before publishing it for review

Care to link it?

Reply:
Leaving this comment out in #4194.